### PR TITLE
Feature/143 display scenarios on screen

### LINKED ISSUE #2: rspec coverage
We need tests for basic level of coverage for:
 - [x] set-up environment to support unit tests
- [x] model specs
- [x] controller specs
- [x] feature specs
- [x] service spec

Notes:
CodeClimate - not sure if it's set-up yet (will craft up a separate story for this - write up a ticket)

### LINKED ISSUE #143: Display YAML tests from DB on screen at /tests rq #142
- [x] setup scenario controller
- [x] scenario index view
- [x] add link to variables for relevant tests
- [x] scenario show view
- [x] show inputs
- [x] show outputs

requires #142 
part of #62

### PR DESCRIPTION
**Pull request by mischa-s**
closes #143 
closes #2

Screenshot_2019-04-30 Rapu Ture - Connecting the Rules(1)
Screenshot_2019-04-30 Rapu Ture - Connecting the Rules

### CODE REVIEW

**Comment by lamorrison** [comment]
What's coffescript for?

File: app/views/scenarios/_properties.html.haml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        - groups.each do |group, residents|
+          = group.humanize
+          %ul
+            - if residents.class == Array
+              - residents.each do |resident|
+                %li
+                  = "#{resident} 👤"
+            - else
+              %li
+                = "#{residents} 👤"
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/_properties.html.haml]: Like this should work
```
#{residents} 👤
```

File: app/services/scenarios_fetch_service.rb [hunk 2]
```diff
@@ -58,14 +58,14 @@ def self.find_all_duplicates(array)
 
       dup << v if map[v] == 2
     end
-    raise StandardError.new("These scenarios have duplicate names: #{dups} !!!!!") if dup[0]
+    raise StandardError.new("These scenarios have duplicate names: #{dup} !!!!!") if dup[0]
```
Comment (Br3nda) [inline app/services/scenarios_fetch_service.rb]: good catch!

File: app/controllers/scenarios_controller.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class ScenariosController < ApplicationController
+  def index
```
Comment (Br3nda) [inline app/controllers/scenarios_controller.rb]: do you think you can add some test converage for the controller, or maybe a feature spec, so this code is run in tests.

File: app/views/scenarios/show.html.haml [hunk 4]
```diff
@@ -0,0 +1,45 @@
+.container
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables 
+    - elsif key1 == 'families'
+      = render 'families', families: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'families', families: value1, variables: @variables
```

[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]: @mischa-s I worked out how to do suggestions! sorry for all the commits in your branch

File: app/views/scenarios/_outputs.html.haml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+          %h4.card-title
+            = link_to vari.description, vari || name
+          - value1.each do |key, value2|
+            %ul
+              %li
+                %h5
+                  #{key}:
+                  - if value2.class == Array
+                    - value2.each do |value3|
+                      = "#{value3}$ "
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/_outputs.html.haml]:
```suggestion
                      "#{value3}$
```

File: app/views/scenarios/_outputs.html.haml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+                  - else
+                    = "#{value2}$ "
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/_outputs.html.haml]:
```suggestion
                    #{value2}$
```

File: app/views/scenarios/show.html.haml [hunk 7]
```diff
@@ -0,0 +1,44 @@
+.container
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables
+    - elsif key1 == 'families'
+      = render 'families', families: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'families', families: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables
+    - elsif key1 == 'families'
+      = render 'families', families: value1, variables: @variables 
+    - elsif key1 == 'titled_properties'
+      = render 'properties', properties: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'properties', properties: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    - elsif key1 == 'families'
+      = render 'families', families: value1, variables: @variables 
+    - elsif key1 == 'titled_properties'
+      = render 'properties', properties: value1, variables: @variables 
+    - else
+      - if value1.class == Hash || value1.class == Array
+        %hr/
+        %i
+          %h3
+          = "#{key1.humanize}: "
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
          #{key1.humanize}:
```

File: app/views/scenarios/show.html.haml [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        - value1.each do |key2, value2|
+          .card
+            .card-body
+              %h4.card-title
+                = "#{key2.humanize}:"
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
                #{key2.humanize}:
```

File: app/views/scenarios/_properties.html.haml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mischa-s) [reply app/views/scenarios/_properties.html.haml]: nope @Br3nda  that doesnt seem to work, I'm not sure why it doesn't like the string interpolation

**Comment by Br3nda** [comment]
@mischa-s if you add to `percy_spec.rb` we'll get screen shots of the pages

File: app/views/scenarios/show.html.haml [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+              - if value2.class == Hash || value2.class == Array
+                - value2.each do |key3, value3|
+                  = "#{key3.humanize}: #{value3}"
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
                  #{key3.humanize}: #{value3}
```

File: app/views/scenarios/show.html.haml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'families', families: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml [hunk 12]
```diff
@@ -0,0 +1,45 @@
+.container
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
    = render 'persons', people: value1, variables: @variables
```

File: app/controllers/scenarios_controller.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
...
+    @scenarios = Scenario.all
+  end
+
+  def show
+    @scenario = Scenario.find(params[:id])
```
Comment (Br3nda) [inline app/controllers/scenarios_controller.rb]: can we have a test for #show too?

Comment (Br3nda) [reply app/controllers/scenarios_controller.rb]: a super quick way to add coverage is just adding to percy_spec.rb

**Comment by Br3nda** [comment]
I don't quite get how it all hangs together, but i've made some scenarios in my dev database and seen them display.

File: app/views/scenarios/show.html.haml [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables 
+    - elsif key1 == 'families'
+      = render 'families', families: value1, variables: @variables 
+    - elsif key1 == 'titled_properties'
+      = render 'properties', properties: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggesting
      = render 'properties', properties: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml [hunk 15]
```diff
@@ -0,0 +1,44 @@
+.container
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables 
+    - elsif key1 == 'families'
+      = render 'families', families: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'families', families: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml [hunk 16]
```diff
@@ -0,0 +1,44 @@
+.container
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'persons', people: value1, variables: @variables
```
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'persons', people: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml (hunk 14 above)
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
       = render 'properties', properties: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml (hunk 15 above)
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
      = render 'families', families: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml (hunk 16 above)
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
       = render 'properties', properties: value1, variables: @variables
```

[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/show.html.haml]:
```suggestion
 = render 'properties', properties: value1, variables: @variables
```

File: app/views/scenarios/show.html.haml [hunk 17]
```diff
@@ -0,0 +1,44 @@
+.container
+  %h1= @scenario.name
+  = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+       = render 'properties', properties: value1, variables: @variables
```
Comment (Br3nda) [inline app/views/scenarios/show.html.haml]:
```suggestion
      = render 'properties', properties: value1, variables: @variables
```

File: app/views/scenarios/_properties.html.haml [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+      - if groups.is_a? Enumerable
+        - groups.each do |group, residents|
+          = group.humanize
+          %ul
+            - if residents.class == Array
+              - residents.each do |resident|
+                %li
+                  #{resident} 👤
+            - else
+              %li
```
Comment (Br3nda) [inline app/views/scenarios/_properties.html.haml]:
```suggestion
              %li #{residents} 👤
```

File: app/views/scenarios/_properties.html.haml [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+                #{residents} 👤
```
Comment (Br3nda) [inline app/views/scenarios/_properties.html.haml]:
```suggestion
```

File: app/views/scenarios/_persons.html.haml [hunk 20]
```diff
@@ -0,0 +1,19 @@
...
+        - attributes.each do |name, value1|
+          - v = variables.select { |variable| variable['name'] == name }[0]
+          %h5
+            = link_to v.description, v
+            - if value1.class == Hash
+              - value1.map do |key2, value2|
+                %h5.li
+                  #{key2}: #{value2}
+            - else
+              ~ #{value1}
```
Comment (Br3nda) [inline app/views/scenarios/_persons.html.haml]: haml is interpreting this as a comment

File: app/views/scenarios/show.html.haml [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+              - else
+                = value2
+      - else
+        - vari = @variables.select { |variable| variable['name'] == key1 }[0]
+        %hr/
+        .card
+          .card-body
+            %h6.card-title
+              = link_to vari.description, vari || key
+              ~ #{value1}
```
Comment (Br3nda) [inline app/views/scenarios/show.html.haml]: does the `~` do something?

File: app/views/scenarios/_persons.html.haml (hunk 20 above)
[the remark this answers is not in the document]
Comment (Br3nda) [reply app/views/scenarios/_persons.html.haml]:
```suggestion
```

File: spec/features/percy_spec.rb [hunk 22]
```diff
@@ -39,4 +40,12 @@
     visit variable_path(variable)
     Percy.snapshot(page, name: 'variables#show')
   end
+  it 'scenarios#index' do
+    visit scenarios_path
+    Percy.snapshot(page, name: 'scenarios#index')
+  end
+  it 'scenarios#show' do
+    visit variable_path(variable)
```
Comment (Br3nda) [inline spec/features/percy_spec.rb]: this is the variables#show page.

Comment (Br3nda) [reply spec/features/percy_spec.rb]:
```suggestion
    scenario = FactoryBot.create :scenario
    visit scenario_path(scenario)
```

**Comment by Br3nda** [comment]
I've removed coffee ☕

File: app/views/scenarios/show.html.haml [hunk 23]
```diff
@@ -0,0 +1,39 @@
+.container
+  %h1= @scenario.name
+  - if @scenario.outputs.present?
+    = render 'outputs', variables: @variables
+  %h2 Inputs
+  - @scenario.inputs.each do |key1, value1|
+    - if key1 == 'persons'
+      = render 'properties', properties: value1, variables: @variables
```
Comment (Br3nda) [inline app/views/scenarios/show.html.haml]: ``suggestion
      = render 'persons', people: value1, variables: @variables
```

**Pull request merged**